File: express/src/api/routes/helloWorld.ts
import { Router, Request, Response, NextFunction } from "express";
import { z } from "../../models/schemas";
import { Authenticated, Admin } from "../middlewares";
import { swaggerRegistery } from "../openapi";
import { sendNewsletter } from "../../worker/tasks/email";

export const helloWorldRouter = Router();

// Hello World Endpoint
async function hello(req: Request, res: Response, next: NextFunction) {
    sendNewsletter("Slim", "mslimbeji@gmail.com");
    res.status(200).json({ message: "Hello World!" });
}

helloWorldRouter.get("/", hello);

swaggerRegistery.registerPath({
    method: "get",
    path: "/hello-world",
    responses: {
        200: {
            description: "A simple response with message field",
            content: {
                "application/json": {
                    schema: z.object({
                        message: z.string().openapi({
                            example: "Hello World",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Hello World"],
    summary: "Hello World Endpoint",
});

// Authenticated Hello World
async function helloUser(req: Request, res: Response, next: NextFunction) {
    res.status(200).json({ message: `Hello ${req.currentUser?.name}!` });
}

helloWorldRouter.get("/user", Authenticated, helloUser);

swaggerRegistery.registerPath({
    method: "get",
    path: "/hello-world/user",
    responses: {
        200: {
            description: "A simple response with message field",
            content: {
                "application/json": {
                    schema: z.object({
                        message: z.string().openapi({
                            example: "Hello Slim",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Hello World"],
    summary: "Hello World Endpoint for authenticated users",
    security: [
        {
            BearerAuth: [],
        },
    ],
});

// Hello World for admins
async function helloAdmin(req: Request, res: Response, next: NextFunction) {
    res.status(200).json({
        message: `Hello Admin ${req.currentUser?.name}!`,
    });
}

helloWorldRouter.get("/admin", Admin, helloAdmin);

swaggerRegistery.registerPath({
    method: "get",
    path: "/hello-world/admin",
    responses: {
        200: {
            description: "A simple response with message field",
            content: {
                "application/json": {
                    schema: z.object({
                        message: z.string().openapi({
                            example: "Hello Admin Slim",
                        }),
                    }),
                },
            },
        },
    },
    tags: ["Hello World"],
    summary: "Hello World Endpoint for admins only",
    security: [
        {
            BearerAuth: [],
        },
    ],
});

File: backend-gin/internal/api/routes/hello_world.go
package routes

import (
	"backend/internal/api/middlewares"
	"backend/internal/background/publishers"
	"backend/internal/lib/gin_"
	"backend/internal/models/schemas"
	"fmt"
	"net/http"

	"github.com/gin-gonic/gin"
)

type HelloResponse struct {
	Message string `json:"message" example:"Hello World!"`
}

// @Summary      Hello World Endpoint
// @Tags         Hello World
// @Produce      json
// @Success      200  {object}  HelloResponse
// @Router       /api/hello-world/ [get]
func hello(c *gin.Context) {
	_, err := publishers.SendNewsletter("Slim", "mslimbeji@gmail.com")
	if err != nil {
		gin_.AbortWithStatusJSON(c, err)
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Hello World!",
	})
}

type HelloUserResponse struct {
	Message string `json:"message" example:"Hello Slim"`
}

// @Summary      Hello World Endpoint for authenticated users
// @Tags         Hello World
// @Produce      json
// @Security     OAuth2Password[admin]
// @Success      200  {object}  HelloUserResponse
// @Router       /api/hello-world/user [get]
func helloUser(c *gin.Context) {
	user := c.MustGet("currentUser").(schemas.UserRead)
	message := fmt.Sprintf("Hello %s!", user.Name)
	c.JSON(http.StatusOK, gin.H{
		"message": message,
	})
}

type HelloAdminResponse struct {
	Message string `json:"message" example:"Hello Admin Slim"`
}

// @Summary      Hello World Endpoint for admins only
// @Tags         Hello World
// @Produce      json
// @Security     OAuth2Password[admin]
// @Success      200  {object}  HelloAdminResponse
// @Router       /api/hello-world/admin [get]
func helloAdmin(c *gin.Context) {
	user := c.MustGet("currentUser").(schemas.UserRead)
	message := fmt.Sprintf("Hello Admin %s!", user.Name)
	c.JSON(http.StatusOK, gin.H{
		"message": message,
	})
}

func RegisterHelloWorld(r *gin.Engine) {
	router := r.Group("/api/hello-world")
	router.GET("", hello)
	router.GET("/user", middlewares.Authenticated, helloUser)
	router.GET("/admin", middlewares.Admin, helloAdmin)
}
